### PR TITLE
Add filtered RBFs and repulsion 

### PR DESCRIPTION
**Pull request by sayeg84**
# PR Checklist

- [] Bug fix
- [x] Feature addition/change
- [] Documentation addition/change
- [] Test addition/change
- [] Black formatting

---

### Describe your changes here:

This introduces Filtered RBFs which can be regularized, a new exponential repulsion, and a version for both power law and exponential repulsion that can be used with a cutoff.

### CODE REVIEW

File: src/mlcg/nn/prior/repulsion.py [hunk 1]
```diff
@@ -236,3 +236,255 @@ def neighbor_list(topology: Topology) -> Dict:
...
+            populated with the predicted energies for each
+            example/structure
+        """
+
+        features = self.data2features(data)
+        # filter all distances that are larger than the cutoff
+        mask = features < self.cutoff
+        features = features[mask]
+        mapping = data.neighbor_list[self.name]["index_mapping"][:, mask]
+        mapping_batch = data.neighbor_list[self.name]["mapping_batch"][mask]
```
Comment (jacopoventurin) [inline src/mlcg/nn/prior/repulsion.py:282]: I see that this make sense in the context of small number of atoms but it probably can become very inefficient if we go to larger systems. Does it make sense to think about using the same NL from the model or recomputing it on the fly here?

File: src/mlcg/nn/prior/repulsion.py [hunk 2]
```diff
@@ -236,3 +236,255 @@ def neighbor_list(topology: Topology) -> Dict:
...
```
Comment (jacopoventurin) [inline src/mlcg/nn/prior/repulsion.py:462]: Same as previous comment:  `I see that this make sense in the context of small number of atoms but it probably can become very inefficient if we go to larger systems. Does it make sense to think about using the same NL from the model or recomputing it on the fly here?`

File: src/mlcg/nn/interpretable_models.py [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+                self.max_num_neighbors,
+                exclude_pair_indices=data.get("exc_pair_index"),
+            )
+        else:
+            edge_index = neighbor_list["index_mapping"]
+            distances = compute_distances(
+                data.pos,
+                edge_index,
+                neighbor_list["cell_shifts"],
+            )
```
Comment (jacopoventurin) [inline src/mlcg/nn/interpretable_models.py]: Here the NL computation is not consistent with other models where different `nls_distance_method` are allowed and set as input parameters. I would make it consistent.

File: src/mlcg/nn/interpretable_models.py [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+    ) -> torch.Tensor:
+        return x_j * (torch.abs(edge_params.view(-1, 1)) * W)
+
+    def propagate(self, edge_index, size=None, **kwargs):
+        # Manually control the flow of arguments
+        kwargs["edge_params"] = kwargs.get("edge_params")
+        return super().propagate(edge_index, size=size, **kwargs)
+
+
+class EdgeRBFRegularizedSchNet(StandardSchNet):
```
Comment (jacopoventurin) [inline src/mlcg/nn/interpretable_models.py]: There is no unit or integration test for this class, probably would be beneficial to have at least one unit test on the same style as the one done for `SchNet` and other models

File: src/mlcg/nn/radial_basis/exp_normal.py [hunk 5]
```diff
@@ -124,6 +124,64 @@ def forward(self, dist: torch.Tensor) -> torch.Tensor:
         )
 
 
+class FilteredExpNormalBasis(ExpNormalBasis):
```
Comment (jacopoventurin) [inline src/mlcg/nn/radial_basis/exp_normal.py]: Unless there is a necessity to have a class that has as attribute the original lambda filter mask, I don't see the point for keeping an extra class like this. 
Probably would be easyer to maintain if we define a sort of converter from `ExpNormalBasis` to another `ExpNormalBasis` with internal parameters pruned accoarding to the provided `lambda` parameters, avoiding having to deal with too many copyes of the same class. This approach can also easily be extended on different RBF classes with the same converter having different set of parameters to be modified for different classes. 

Another approach  could be define `@classmethod` directly in the main class that prune the initialized rbf given the `lambda` parameters (same as done for example in `CutoffExpRepulsion`)

File: src/mlcg/nn/radial_basis/gaussian.py [hunk 6]
```diff
@@ -100,3 +100,57 @@ def forward(self, dist: torch.Tensor) -> torch.Tensor:
             self.coeff * torch.pow(dist - self.offset, 2)
         ) * self.cutoff(dist)
         return expanded_distances
+
+
+class FilteredGaussianBasis(GaussianBasis):
```
Comment (jacopoventurin) [inline src/mlcg/nn/radial_basis/gaussian.py]: Same comment done for `FilteredExpNormalBasis`

File: src/mlcg/nn/radial_basis/regularized_basis.py [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+
+    def forward(self, i, j):
+        r"""
+        Given indices i,j return the corresponding symmetric tensor entry
+        of shape (i.shape, size).
+        Here i,j are indeces tensors of same shape.
+        """
+        i_, j_ = torch.minimum(i, j), torch.maximum(i, j)
+        k = ((2 * self.N - i_ + 1) * i_) // 2 + (j_ - i_)
+        return self.w[k]
```
Comment (jacopoventurin) [inline src/mlcg/nn/radial_basis/regularized_basis.py]: Here there is not guard against i,j values larger than N. This might lead to silently wrong results. Probably worth thinking on how to find an efficient check for that

File: src/mlcg/nn/radial_basis/regularized_basis.py [hunk 8]
```diff
@@ -0,0 +1,254 @@
...
+
+    def plot(self, i, j, layer=0, ax=None, **kwargs):
+        r"""Method for quickly visualizing a specific basis. This is useful for
+        inspecting the distance coverage of basis functions for non-default lower
+        and upper cutoffs.
+
+        Parameters:
+        ----------
+        distances : torch.Tensor
+            Tensor of shape (num_edges,) containing the distances between atom pairs.
```
Comment (jacopoventurin) [inline src/mlcg/nn/radial_basis/regularized_basis.py]: wrong docs: no distances needed in input

File: src/mlcg/nn/losses.py [hunk 9]
```diff
@@ -181,6 +181,61 @@ def forward(self, data: AtomicData) -> torch.Tensor:
...
+        if self.reg_kwd not in data.out:
+            raise RuntimeError(
+                f"target property {self.reg_kwd} has not been computed in data.out {list(data.out.keys())}"
+            )
+
+        return self.reg_factor * torch.nn.functional.l1_loss(
+            data.out[self.reg_kwd],
+            torch.zeros_like(data.out[self.reg_kwd]),
+            reduction=self.reduction,
+        )
```
Comment (jacopoventurin) [inline src/mlcg/nn/losses.py]: Do we need to specify a `reg_factor` here? Since we use by default `Loss` this can be specified by the `weights` parameter. It is also safer when different losses are used to be sure the relative weight between them is correct. Furthermore `weights` can be modified during the training if needed by using the `LossScheduler` callback.

File: src/mlcg/nn/schnet.py [hunk 10]
```diff
@@ -661,3 +663,149 @@ def __init__(
...
+        if use_custom_kernel:
+            distances, edge_index = radius_distance(
+                data.pos,
+                self.rbf_layer.cutoff.cutoff_upper,
+                data.batch,
+                False,  # no loop edges due to compatibility & backward breaks with zero distance
+                self.max_num_neighbors,
+                exclude_pair_indices=data.get("exc_pair_index"),
+            )
+        else:
```
Comment (jacopoventurin) [inline src/mlcg/nn/schnet.py]: As before, this part should be consistent with the new NL computation approach

**Review by jacopoventurin — COMMENTED** [review]
Thank you so much for this PR, I left some comments.

File: src/mlcg/nn/prior/repulsion.py (hunk 1 above)
Comment (sayeg84) [reply src/mlcg/nn/prior/repulsion.py:282]: I agree that this is not an efficient implementation but it is at least as fast as the existint repulsions, and for testing could work. 

We can't use the same NL from the model because the repulsion has some exclusions of residues. We would need to implement some filter to remove this excluded residues. How to implement this filter is not trivial because it would need to be both backwards compatible and fast, and the existing neighbor lists are quite large so we can't just search over them, even with a hash, it would be too slow. 

The efficient implementation needs more careful planning but maybe its good to start rolling this out.

File: src/mlcg/nn/interpretable_models.py (hunk 3 above)
Comment (Jamba15) [reply src/mlcg/nn/interpretable_models.py]: I will!

File: src/mlcg/nn/losses.py (hunk 9 above)
Comment (jacopoventurin) [reply src/mlcg/nn/losses.py]: removed

File: src/mlcg/nn/radial_basis/exp_normal.py (hunk 5 above)
Comment (jacopoventurin) [reply src/mlcg/nn/radial_basis/exp_normal.py]: done

File: src/mlcg/nn/radial_basis/gaussian.py (hunk 6 above)
Comment (jacopoventurin) [reply src/mlcg/nn/radial_basis/gaussian.py]: done

File: src/mlcg/nn/radial_basis/regularized_basis.py (hunk 7 above)
Comment (jacopoventurin) [reply src/mlcg/nn/radial_basis/regularized_basis.py]: done

File: src/mlcg/nn/radial_basis/regularized_basis.py (hunk 8 above)
Comment (jacopoventurin) [reply src/mlcg/nn/radial_basis/regularized_basis.py]: fixed

File: src/mlcg/nn/schnet.py (hunk 10 above)
Comment (jacopoventurin) [reply src/mlcg/nn/schnet.py]: done

**Review by jacopoventurin — APPROVED** [review]
